### PR TITLE
[Refactor] refactor apis for HDFS write.

### PR DESCRIPTION
**Pull request by FMX**
# [BUG]/[FEATURE] title

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

**Comment by FMX** [comment]
ping @waitinfuture @AngersZhuuuu

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/Writer.java [hunk 1]
```diff
@@ -392,4 +398,20 @@ public long getSplitThreshold() {
   public PartitionSplitMode getSplitMode() {
     return splitMode;
   }
+
```
Comment (FMX) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/Writer.java:401]: This section is due to scala 2.11 restriction, the trait can not be treated as an interface with default methods.

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala [hunk 2]
```diff
@@ -36,36 +36,57 @@ import io.netty.buffer.{CompositeByteBuf, Unpooled}
...
+
+  def notifySlowFlush(dirs: ListBuffer[File]): Unit = {}
+
+  def reportError(workingDir: mutable.Buffer[File], e: IOException,
+    deviceErrorType: DeviceErrorType): Unit = {}
+}
+
+private[worker] abstract class FlushTask(
+  val buffer: CompositeByteBuf,
+  val notifier: FlushNotifier) {
```
Comment (FMX) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala:63]: HDFS does not support NIO, so here needs some modification.

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala [hunk 3]
```diff
@@ -324,9 +362,9 @@ private[worker] final class LocalStorageManager(
         }
```
Comment (AngersZhuuuu) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala:362]: Should we change `LocalStorageManager` to `StorageManager`?
Or we will implement different `StorageManager` for each storage?

Comment (FMX) [reply server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala:362]: Yes, you are right. I should change it to `StorageManager`.

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/Writer.java [hunk 4]
```diff
@@ -47,8 +48,8 @@
  * Note: Once FlushNotifier.exception is set, the whole file is not available.
  *       That's fine some of the internal state(e.g. bytesFlushed) may be inaccurate.
  */
-public final class FileWriter extends DeviceObserver {
-  private static final Logger logger = LoggerFactory.getLogger(FileWriter.class);
+public final class Writer implements DeviceObserver {
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/Writer.java:51]: I think FileWriter is fine, hdfs files and oss files are files, anyway.

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/StorageManager.scala [hunk 5]
```diff
@@ -36,36 +36,53 @@ import io.netty.buffer.{CompositeByteBuf, Unpooled}
...
-  private val workingQueues = new Array[LinkedBlockingQueue[FlushTask]](threadCount)
-  private val bufferQueue = new LinkedBlockingQueue[CompositeByteBuf]()
-  private val workers = new Array[Thread](threadCount)
-  private var nextWorkerIndex: Int = 0
-  private val flushCount = new LongAdder
-  private val flushTotalTime = new LongAdder
-  private val avgTimeWindow = new Array[(Long, Long)](flushAvgTimeWindowSize)
-  private var avgTimeWindowCurrentIndex = 0
+    val flushAvgTimeMinimumCount: Int) extends Logging {
+  protected lazy val diskFlusherId = System.identityHashCode(this)
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/StorageManager.scala:77]: flusherId

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/StorageManager.scala [hunk 6]
```diff
@@ -83,22 +100,20 @@ private[worker] final class Flusher(
         override def run(): Unit = {
           while (!stopFlag.get()) {
             val task = workingQueues(index).take()
-            val key = s"Flusher-$workingDirs-${rand.nextInt()}"
+            val key = s"LocalFlusher-$this-${rand.nextInt()}"
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/StorageManager.scala:103]: We don't know it's local or hdfs/oss in parent class here

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/StorageManager.scala [hunk 7]
```diff
@@ -36,36 +36,53 @@ import io.netty.buffer.{CompositeByteBuf, Unpooled}
...
+    val notifier: FlushNotifier) {
+  def flush(): Unit
+}
 
-private[worker] final class Flusher(
-    val workingDirs: mutable.Buffer[File],
-    workerSource: AbstractSource,
-    val deviceMonitor: DeviceMonitor,
+private[worker] class LocalFlushTask(
+    override val buffer: CompositeByteBuf,
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/StorageManager.scala:64]: remove "override val" and other places

File: common/src/main/java/com/aliyun/emr/rss/common/protocol/StorageInfo.java [hunk 8]
```diff
@@ -37,6 +37,11 @@ public int getType() {
   public StorageInfo() {
   }
 
+  public StorageInfo(Type type, boolean isFinal) {
```
[the remark this answers is not in the document]
Comment (FMX) [reply common/src/main/java/com/aliyun/emr/rss/common/protocol/StorageInfo.java:40]: I think we can do it in later pr and convert Java enumerations to Scala.

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/Writer.java (hunk 4 above)
Comment (FMX) [reply server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/Writer.java:51]: ok

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Pull request merged**